Expose JS API for PPX rewriting

astrada opened the pull request:
Expose a `jsxv2` object with a `rewrite` method to transform OCaml+JSX to pure OCaml, that can be used by JS playgrounds.

Example:
```js
jsxv2.rewrite('(button ~children:[ReasonReact.stringToElement "Hello!"] ()) [@JSX ]')
```
Returns:
```ocaml
let _ =
  ReactDOMRe.createElement "button"
    [|(ReasonReact.stringToElement "Hello!")|]
```
If there are no errors in the input string, the output code is returned in a serialized JSON object, e.g.:
```json
{"ocaml_code": "let _ =\n  ReactDOMRe.createElement \"button\"\n    [|(ReasonReact.stringToElement \"Hello!\")|]"}
```
Otherwise, you will get a serialized JSON object containing the error returned by the PPX preprocessor, e.g.:
```js
jsxv2.rewrite('(button () /) [@JSX]')
```
Returns:
```json
{ "ppx_error_msg": "Line 1, 12: Syntax error: operator expected.", "row": 0, "column": 12, "endRow": 0, "endColumn": 13, "text": "Syntax error: operator expected.", "type": "error" }
```

Signed-off-by: Alessandro Strada <alessandro.strada@gmail.com>

astrada commented:
I've opened a companion PR (facebook/reason#1526) to align `reactjs_jsx_ppx_v2.ml`.

chenglou commented:
Sorry a little busy. Will test this internally soon. Ping me if I don't review by Monday. Thanks!

astrada commented:
Sure! No problem, thanks!

chenglou commented on jscomp/bin/reactjs_jsx_ppx_v2.bspp.ml:
```diff
@@ -312,7 +312,75 @@ let jsxMapper () =
+    | None ->
+        Printf.sprintf "{\"js_error_msg\": %S}" (Printexc.to_string e)
+
+module Js = struct
```
@bobzhang are you ok with this here? Seems copied over from jsoo

astrada replied:
Yes, it's based on https://github.com/BuckleScript/bucklescript/blob/master/jscomp/core/jsoo_main.ml#L49 (to provide a consistent API). Let me know if I have to modify it.

chenglou commented:
Well it compiles and isn't invoked until you need it, so I think that's fine! One last check: have you tried it in the playground?

astrada commented:
> Well it compiles and isn't invoked until you need it, so I think that's fine!

Thanks!

> One last check: have you tried it in the playground?

Yes, I did. I forked [bucklescript-playground](https://github.com/astrada/bucklescript-playground/tree/jsoo-jsx), modified [compiler_test.js](https://github.com/astrada/bucklescript-playground/blob/jsoo-jsx/compiler_test.js) and added 2 tests: [reason_react.ml](https://github.com/astrada/bucklescript-playground/blob/jsoo-jsx/examples/reason_react.ml) and [reason_react.jsx](https://github.com/astrada/bucklescript-playground/blob/jsoo-jsx/examples/reason_react.jsx). I didn't open a pull request, because these tests require `reason-react` and `bs-platform` to produce `.cmi`s and `.cmj`s and I didn't want to complicate `bucklescript-playground` maintenance.

chenglou commented:
Looks good, thanks! Can you fix the conflicts? I'll merge after that

astrada commented:
OK, done.

chenglou commented:
Thanks! Merging. cc @bobzhang
